### PR TITLE
Improve Polymorphism of PackageReferences

### PR DESCRIPTION
**Pull request by theunrepentantgeek**
**What this PR does / why we need it**:

Removes most of the explicit type downcasting from `PackageReference` to implementation types, mostly as a prepositioning step for the introduction of SubPackageReference.

**Special notes for your reviewer**:

Also fixes up the import aliases we're generating - the intent was to use the simplest form of imports (e.g. `v20230701`) for the current CRD version, and more verbose variations (e.g. `v1beta20230701`) for older CRD versions, but this wasn't updated when we switched to `v1api` based resources.

**How does this PR make you feel**:
gif

**If applicable**:
- [x] this PR contains tests

### CODE REVIEW

File: v2/tools/generator/internal/astmodel/local_package_reference.go [hunk 1]
```diff
@@ -138,6 +138,55 @@ func (pr LocalPackageReference) GroupVersion() (string, string) {
...
+
+var generatorVersionSimplifications = map[string]string{
+	"v1alpha1api": "alpha",
+	"v1beta1api":  "beta",
+	"v1api":       "v",
+}
+
+func (pr LocalPackageReference) simplify(result string, simplifications map[string]string) string {
+	for l, s := range simplifications {
+		result = strings.Replace(result, l, s, -1)
```
Comment (super-harsh) [inline v2/tools/generator/internal/astmodel/local_package_reference.go:185]: Consider using a generic `strings.NewReplacer` here for both `apiVersionSimplifications` and `generatorVersionSimplifications`

Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/local_package_reference.go]: That's really nice, thanks for the pointer. Changed.

File: v2/tools/generator/internal/astmodel/package_reference.go [hunk 2]
```diff
@@ -33,6 +33,8 @@ type PackageReference interface {
 	TryGroupVersion() (string, string, bool)
 	// GroupVersion returns the group and version of this reference, triggering a panic if they aren't available
 	GroupVersion() (string, string)
+	// ImportAlias returns the import alias to use for this package reference
+	ImportAlias(style PackageImportStyle) string
```
[the remark this answers is not in the document]
Comment (theunrepentantgeek) [reply v2/tools/generator/internal/astmodel/package_reference.go:37]: I agree with you that it's not ideal to have `ExternalPackageReference` implementing this. I'm expecting to relocate this method shortly, in an upcoming PR.

**Comment by theunrepentantgeek** [comment]
@microsoft-github-policy-service rerun

**Review by super-harsh — APPROVED** [review]

**Pull request merged**